File: packages/java/oauth/src/main/java/do_/oauth/DeviceFlow.java
package do_.oauth;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import java.io.*;
import java.net.*;
import java.net.http.*;
import java.time.Duration;
import java.util.concurrent.*;
import java.util.function.Consumer;

/**
 * Device flow authorization for OAuth.
 * Implements the device authorization grant flow.
 */
public class DeviceFlow {
    private static final String AUTH_URL = "https://auth.apis.do/user_management/authorize_device";
    private static final String TOKEN_URL = "https://auth.apis.do/user_management/authenticate";
    private static final Gson gson = new Gson();

    private final String clientId;
    private final HttpClient httpClient;

    /**
     * Device authorization response.
     */
    public static class DeviceAuthResponse {
        public String deviceCode;
        public String userCode;
        public String verificationUri;
        public String verificationUriComplete;
        public int expiresIn;
        public int interval;

        @Override
        public String toString() {
            return String.format(
                "Visit %s and enter code: %s",
                verificationUri, userCode
            );
        }
    }

    /**
     * Token response from authentication.
     */
    public static class TokenResponse {
        public String accessToken;
        public String refreshToken;
        public String tokenType;
        public int expiresIn;
        public String error;
        public String errorDescription;

        public boolean isError() {
            return error != null;
        }

        public boolean isPending() {
            return "authorization_pending".equals(error) || "slow_down".equals(error);
        }
    }

    /**
     * Creates a DeviceFlow with the default client ID.
     */
    public DeviceFlow() {
        this("client_01JQYTRXK9ZPD8JPJTKDCRB656");
    }

    /**
     * Creates a DeviceFlow with a custom client ID.
     * @param clientId OAuth client ID
     */
    public DeviceFlow(String clientId) {
        this.clientId = clientId;
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(30))
            .build();
    }

    /**
     * Initiates device authorization.
     * @return Device authorization response
     * @throws IOException If the request fails
     * @throws InterruptedException If the request is interrupted
     */
    public DeviceAuthResponse authorize() throws IOException, InterruptedException {
        String body = String.format("client_id=%s", URLEncoder.encode(clientId, "UTF-8"));

        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(AUTH_URL))
            .header("Content-Type", "application/x-www-form-urlencoded")
            .POST(HttpRequest.BodyPublishers.ofString(body))
            .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() != 200) {
            throw new IOException("Authorization failed: " + response.body());
        }

        JsonObject json = gson.fromJson(response.body(), JsonObject.class);
        DeviceAuthResponse auth = new DeviceAuthResponse();
        auth.deviceCode = json.get("device_code").getAsString();
        auth.userCode = json.get("user_code").getAsString();
        auth.verificationUri = json.get("verification_uri").getAsString();
        if (json.has("verification_uri_complete")) {
            auth.verificationUriComplete = json.get("verification_uri_complete").getAsString();
        }
        auth.expiresIn = json.has("expires_in") ? json.get("expires_in").getAsInt() : 900;
        auth.interval = json.has("interval") ? json.get("interval").getAsInt() : 5;

        return auth;
    }

    /**
     * Polls for token after user authorization.
     * @param deviceCode Device code from authorization response
     * @param interval Polling interval in seconds
     * @param timeout Maximum time to wait in seconds
     * @return Token response
     * @throws IOException If the request fails
     * @throws InterruptedException If polling is interrupted
     * @throws TimeoutException If polling times out
     */
    public TokenResponse pollForToken(String deviceCode, int interval, int timeout)
            throws IOException, InterruptedException, TimeoutException {
        return pollForToken(deviceCode, interval, timeout, null);
    }

    /**
     * Polls for token after user authorization with progress callback.
     * @param deviceCode Device code from authorization response
     * @param interval Polling interval in seconds
     * @param timeout Maximum time to wait in seconds
     * @param onPoll Callback for each poll attempt (optional)
     * @return Token response
     * @throws IOException If the request fails
     * @throws InterruptedException If polling is interrupted
     * @throws TimeoutException If polling times out
     */
    public TokenResponse pollForToken(String deviceCode, int interval, int timeout,
            Consumer<Integer> onPoll) throws IOException, InterruptedException, TimeoutException {
        long startTime = System.currentTimeMillis();
        long timeoutMs = timeout * 1000L;
        int pollInterval = interval;
        int attempts = 0;

        while (System.currentTimeMillis() - startTime < timeoutMs) {
            attempts++;
            if (onPoll != null) {
                onPoll.accept(attempts);
            }

            String body = String.format(
                "client_id=%s&device_code=%s&grant_type=urn:ietf:params:oauth:grant-type:device_code",
                URLEncoder.encode(clientId, "UTF-8"),
                URLEncoder.encode(deviceCode, "UTF-8")
            );

            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(TOKEN_URL))
                .header("Content-Type", "application/x-www-form-urlencoded")
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            JsonObject json = gson.fromJson(response.body(), JsonObject.class);

            TokenResponse token = new TokenResponse();

            if (json.has("error")) {
                token.error = json.get("error").getAsString();
                if (json.has("error_description")) {
                    token.errorDescription = json.get("error_description").getAsString();
                }

                if ("slow_down".equals(token.error)) {
                    pollInterval += 5;
                }

                if (!token.isPending()) {
                    return token;
                }
            } else {
                token.accessToken = json.get("access_token").getAsString();
                if (json.has("refresh_token")) {
                    token.refreshToken = json.get("refresh_token").getAsString();
                }
                token.tokenType = json.has("token_type") ? json.get("token_type").getAsString() : "Bearer";
                token.expiresIn = json.has("expires_in") ? json.get("expires_in").getAsInt() : 3600;
                return token;
            }

            Thread.sleep(pollInterval * 1000L);
        }

        throw new TimeoutException("Authorization timed out after " + timeout + " seconds");
    }

    /**
     * Gets the client ID.
     * @return OAuth client ID
     */
    public String getClientId() {
        return clientId;
    }
}

File: packages/kotlin/oauth/src/main/kotlin/do_/oauth/DeviceFlow.kt
package do_.oauth

import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.request.*
import io.ktor.client.request.forms.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.coroutines.delay
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.util.concurrent.TimeoutException

/**
 * Device flow authorization for OAuth.
 * Implements the device authorization grant flow.
 */
class DeviceFlow(
    private val clientId: String = DEFAULT_CLIENT_ID
) {
    companion object {
        private const val AUTH_URL = "https://auth.apis.do/user_management/authorize_device"
        private const val TOKEN_URL = "https://auth.apis.do/user_management/authenticate"
        const val DEFAULT_CLIENT_ID = "client_01JQYTRXK9ZPD8JPJTKDCRB656"
    }

    private val json = Json { ignoreUnknownKeys = true }
    private val httpClient = HttpClient(CIO) {
        engine {
            requestTimeout = 30_000
        }
    }

    /**
     * Device authorization response.
     */
    @Serializable
    data class DeviceAuthResponse(
        val deviceCode: String,
        val userCode: String,
        val verificationUri: String,
        val verificationUriComplete: String? = null,
        val expiresIn: Int = 900,
        val interval: Int = 5
    ) {
        override fun toString(): String = "Visit $verificationUri and enter code: $userCode"
    }

    /**
     * Token response from authentication.
     */
    @Serializable
    data class TokenResponse(
        val accessToken: String? = null,
        val refreshToken: String? = null,
        val tokenType: String = "Bearer",
        val expiresIn: Int = 3600,
        val error: String? = null,
        val errorDescription: String? = null
    ) {
        fun isError(): Boolean = error != null
        fun isPending(): Boolean = error == "authorization_pending" || error == "slow_down"
    }

    /**
     * Initiates device authorization.
     * @return Device authorization response
     */
    suspend fun authorize(): DeviceAuthResponse {
        val response = httpClient.submitForm(
            url = AUTH_URL,
            formParameters = parameters {
                append("client_id", clientId)
            }
        )

        if (response.status != HttpStatusCode.OK) {
            throw RuntimeException("Authorization failed: ${response.bodyAsText()}")
        }

        val jsonBody = json.parseToJsonElement(response.bodyAsText()).jsonObject
        return DeviceAuthResponse(
            deviceCode = jsonBody["device_code"]?.jsonPrimitive?.content ?: "",
            userCode = jsonBody["user_code"]?.jsonPrimitive?.content ?: "",
            verificationUri = jsonBody["verification_uri"]?.jsonPrimitive?.content ?: "",
            verificationUriComplete = jsonBody["verification_uri_complete"]?.jsonPrimitive?.content,
            expiresIn = jsonBody["expires_in"]?.jsonPrimitive?.content?.toIntOrNull() ?: 900,
            interval = jsonBody["interval"]?.jsonPrimitive?.content?.toIntOrNull() ?: 5
        )
    }

    /**
     * Polls for token after user authorization.
     * @param deviceCode Device code from authorization response
     * @param interval Polling interval in seconds
     * @param timeout Maximum time to wait in seconds
     * @param onPoll Callback for each poll attempt (optional)
     * @return Token response
     */
    suspend fun pollForToken(
        deviceCode: String,
        interval: Int = 5,
        timeout: Int = 900,
        onPoll: ((Int) -> Unit)? = null
    ): TokenResponse {
        val startTime = System.currentTimeMillis()
        val timeoutMs = timeout * 1000L
        var pollInterval = interval
        var attempts = 0

        while (System.currentTimeMillis() - startTime < timeoutMs) {
            attempts++
            onPoll?.invoke(attempts)

            val response = httpClient.submitForm(
                url = TOKEN_URL,
                formParameters = parameters {
                    append("client_id", clientId)
                    append("device_code", deviceCode)
                    append("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
                }
            )

            val jsonBody = json.parseToJsonElement(response.bodyAsText()).jsonObject

            if (jsonBody.containsKey("error")) {
                val error = jsonBody["error"]?.jsonPrimitive?.content
                val errorDescription = jsonBody["error_description"]?.jsonPrimitive?.content

                if (error == "slow_down") {
                    pollInterval += 5
                }

                if (error != "authorization_pending" && error != "slow_down") {
                    return TokenResponse(
                        error = error,
                        errorDescription = errorDescription
                    )
                }
            } else {
                return TokenResponse(
                    accessToken = jsonBody["access_token"]?.jsonPrimitive?.content,
                    refreshToken = jsonBody["refresh_token"]?.jsonPrimitive?.content,
                    tokenType = jsonBody["token_type"]?.jsonPrimitive?.content ?: "Bearer",
                    expiresIn = jsonBody["expires_in"]?.jsonPrimitive?.content?.toIntOrNull() ?: 3600
                )
            }

            delay(pollInterval * 1000L)
        }

        throw TimeoutException("Authorization timed out after $timeout seconds")
    }

    /**
     * Closes the HTTP client.
     */
    fun close() {
        httpClient.close()
    }

    /**
     * Gets the client ID.
     * @return OAuth client ID
     */
    fun getClientId(): String = clientId
}
